NMS-12731: ResourceDao performance enhancements for the Time Series Storage API - Part 3 (#3038)
OpenNMS/opennms: opennms-doc/guide-admin/src/asciidoc/text/operation/timeseries/configuration.adoc

:imagesdir: ../../../images
==== Configuration

> ===== Enabling Time series Integration Layer

Reviewer: ```suggestion
===== Enabling Time Series Integration Layer
```

> _{opennms-product-name}_ can be configured to use the Time series Integration Layer by setting the following property in `${OPENNMS_HOME}/etc/opennms.properties`:

Reviewer: ```suggestion
_{opennms-product-name}_ can be configured to use the Time Series Integration Layer by setting the following property in `${OPENNMS_HOME}/etc/opennms.properties`:
```

[source]
----
org.opennms.timeseries.strategy=integration
----

> After activating the _Time series Integration Layer_, you need to start an actual implementation.

Reviewer: ```suggestion
After activating the _Time Series Integration Layer_, you need to start an actual implementation.
```

For specific instructions, check your plugin description.

> We also highly recommend that you reference resources stored in the _Time series Integration Layer_ by their foreign source and foreign ID, as opposed to their database ID.

Reviewer: ```suggestion
We also highly recommend that you reference resources stored in the _Time Series Integration Layer_ by their foreign source and foreign ID, as opposed to their database ID.
```

To this end, set the following property in the same file:

[source]
----
org.opennms.rrd.storeByForeignSource=true
----

> With these set, _{opennms-product-name}_ will begin persisting metrics using the _Time series Integration Layer_ when restarted.

Reviewer: ```suggestion
With these set, _{opennms-product-name}_ will begin persisting metrics using the _Time Series Integration Layer_ when restarted.
```

Additional configuration options are presented in the next section.

===== Configuration Reference

> The following properties, found in `${OPENNMS_HOME}/etc/opennms.properties`, can be used to configure and tune the _Time series Integration Layer_.

Reviewer: ```suggestion
The following properties, found in `${OPENNMS_HOME}/etc/opennms.properties`, can be used to configure and tune the _Time Series Integration Layer_.
```

[[ga-opennms-operation-timeseries-properties-general]]
====== General
[options="header, autowidth"]
|===
| Name                                            | Default              | Description
| `org.opennms.timeseries.ring_buffer_size`     | `8192`               | Maximum number of records that can be held in the ring buffer. Must be a power of two.

> | `org.opennms.timeseries.config.writer_threads`       | `16`                 | Number of threads used to pull samples from the ring buffer and insert them into the Time series Database.

Reviewer: ```suggestion
| `org.opennms.timeseries.config.writer_threads`       | `16`                 | Number of threads used to pull samples from the ring buffer and insert them into the Time Series Database.
```

The following properties, found in `${OPENNMS_HOME}/etc/opennms.properties`, can be used to configure and tune the _Timeseries Integration Layer_.

[[ga-opennms-operation-timeseries-properties-general]]
====== General
[options="header, autowidth"]
|===
| Name                                            | Default              | Description
| `org.opennms.timeseries.ring_buffer_size`     | `8192`               | Maximum number of records that can be held in the ring buffer. Must be a power of two.

> | `org.opennms.timeseries.config.writer_threads`       | `16`                 | Number of threads used to pull samples from the ring buffer and insert them into the Timeseries Database.

Reviewer: ```suggestion
| `org.opennms.timeseries.config.writer_threads`       | `16`                 | Number of threads used to pull samples from the ring buffer and insert them into the Time Series Database.
```

Author: @Bonrob2 do you suggest to use the two words 'Time Series' instead of the one word Timeseries? If so I think we should change it everywhere?

Reviewer: Yes, two words in prose would be better, but I am aware that it is inconsistent with the rest of the documentation. This is something I'll update with the doc revamp for 27.0. If it's a lot of work right now, just leave it as one word, as long as it's one word consistently. I can note to update it in 27.0.

Author: I can update it in the time series integration layer and make it here consistent.

| `cache.timeseriesPersisterMetaTagCache.recordStats`        | `true`              | Should cache statistics be exposed via JMX for MetaTagCache?

| `cache.timeseriesSearcherCache.expireAfterRead`    | `300`               | Expiry time in seconds for TimeseriesSearcherCache.
| `cache.timeseriesSearcherCache.maximumSize`        | `8192`              | Maximum size for TimeseriesSearcherCache.
| `cache.timeseriesSearcherCache.recordStats`        | `true`              | Should cache statistics be exposed via JMX for TimeseriesSearcherCache?

| `cache.timeseriesMetaDataCache.expireAfterRead`    | `300`               | Expiry time in seconds for TimeseriesMetaDataCache.
| `cache.timeseriesMetaDataCache.maximumSize`        | `8192`              | Maximum size for TimeseriesMetaDataCache.
| `cache.timeseriesMetaDataCache.recordStats`        | `true`              | Should cache statistics be exposed via JMX for TimeseriesMetaDataCache?
|===

[[ga-opennms-operation-timeseries-properties-recommendations]]
===== Recommendations
Caches have been introduced to improve performance.

> You might need to tune the cache settings to suit your needs, see parameters above.

Reviewer: ```suggestion
You might need to tune the cache settings to suit your needs. See parameters above.
```

[options="header, autowidth"]
|===
| Name                             | Description

> | TimeseriesMetaDataCache          | Caches meta data that uses _{opennms-product-name}_  internally.

Reviewer: ```suggestion
| TimeseriesMetaDataCache          | Caches metadata that uses _{opennms-product-name}_  internally.
```

> | TimeseriesSearcherCache          | Caches Metrics by tag to improve the resource lookup.

Reviewer: ```suggestion
| TimeseriesSearcherCache          | Caches metrics by tag to improve the resource lookup.
```

> | TimeseriesPersisterMetaTagCache  | Caches all additionally configured and resolved meta tag values by resource

Reviewer: ```suggestion
| TimeseriesPersisterMetaTagCache  | Caches all additionally configured and resolved meta-tag values by resource.
```

|===

The samples gathered by the collectors are temporarily stored in a ring buffer before they are persisted to the _Timeseries Integration Layer_.
The value of the `ring_buffer_size` should be increased if you expect large peaks of collectors returning at once or latency in persisting these.
However, note that the memory used by the ring buffer is reserved, and larger values may require an increased heap size.

[[ga-opennms-operation-timeseries-properties-meta-tags]]
===== Expose additional meta tags
Metrics that are stored via the timeseries plugin contain the minimal set of tags for _{opennms-product-name}_ to work.
This might not be sufficient if the data is used outside of _{opennms-product-name}_ as well.

> Therefore additional meta tags can be configured via `${OPENNMS_HOME}/etc/opennms.properties`.

Reviewer: ```suggestion
Configure additional meta tags via `${OPENNMS_HOME}/etc/opennms.properties`.
```

| TimeseriesPersisterMetaTagCache  | Caches all additionally configured and resolved meta-tag values by resource.
|===

> The samples gathered by the collectors are temporarily stored in a ring buffer before they are persisted to the _Time series Integration Layer_.

Reviewer: ```suggestion
The samples gathered by the collectors are temporarily stored in a ring buffer before they are persisted to the _Time Series Integration Layer_.
```

`org.opennms.timeseries.tin.metatags.tag.nodelabel=${node:label}`
`org.opennms.timeseries.tin.metatags.tag.sysObjectID=${node:sys-object-id}`

> Categories can be exposed by setting `org.opennms.timeseries.tin.metatags.exposeCategories` to `true`

Reviewer: ```suggestion
Expose categories by setting `org.opennms.timeseries.tin.metatags.exposeCategories` to `true`.
```
